tiledb/sm/rest/curl.cc

```diff
@@ -180,21 +181,108 @@ static int buffer_list_seek_callback(
+ * actually taken care of if that number differs from the number passed to
+ * your callback function, it signals an error condition to the library
+ *
+ * @param res_data points to the delivered data
+ * @param size the size of that data is size multiplied with nmemb
+ * @param count number of bytes actually taken care of
+ * @param userdata the userdata argument is set with CURLOPT_HEADERDATA
+ * @return
+ */
+size_t write_header_callback(
+    void* res_data, size_t size, size_t count, void* userdata) {
+  size_t header_length = size * count;
+  auto* header_buffer = static_cast<char*>(res_data);
+  auto* pmHeader = static_cast<HeaderCbData*>(userdata);
+
+  if (pmHeader->uri.empty()) {
+    LOG_ERROR("Rest components as array_ns and array_uri cannot be empty");
+    return 0;
+  }
+
+  std::string header(header_buffer, header_length);
+  const size_t header_key_end_pos = header.find(": ");
+
+  if (header_key_end_pos == std::string::npos) {
+    LOG_ERROR(
+        "All response headers should be formatted as key:value elements.");
+    return 0;
+  }
+
+  if (header_key_end_pos != std::string::npos) {
+    std::string header_key = header.substr(0, header_key_end_pos);
+    std::transform(
+        header_key.begin(), header_key.end(), header_key.begin(), ::tolower);
+
+    if (header_key == constants::redirection_header_key) {
+      // Fetch the header value. Subtract 2 from the `header_length` to
+      // remove the trailing CR LF ("\r\n") and subtract another 2
+      // for the ": ". Ensure this ends with a null terminator.
+      const std::string header_value =
+          header.substr(
+              header_key_end_pos + 2, header_length - header_key_end_pos - 4) +
+          "\0";
+
+      // Find the http scheme
+      const size_t header_scheme_end_pos = header_value.find("://");
+
+      if (header_scheme_end_pos == std::string::npos) {
+        LOG_ERROR(
+            "The header `location` should have a value that includes the "
+            "scheme in the URI.");
+        return 0;
+      }
+
+      const std::string header_scheme =
+          header_value.substr(0, header_scheme_end_pos);
+
+      // Find the domain
+      const std::string header_value_scheme_excl =
+          header_value.substr(header_scheme_end_pos + 3, header_value.length());
+      const size_t header_domain_end_pos = header_value_scheme_excl.find("/");
+
+      if (header_domain_end_pos == std::string::npos) {
+        LOG_ERROR(
+            "The header `location` should have a value that includes the "
+            "domain in the URI.");
+        return 0;
+      }
+
+      const std::string header_value_domain =
+          header_value_scheme_excl.substr(0, header_domain_end_pos);
+
+      std::unique_lock<std::mutex> rd_lck(*(pmHeader->redirect_uri_map_lock));
+      pmHeader->redirect_uri_map->insert(std::make_pair(
+          pmHeader->uri, header_scheme + "://" + header_value_domain + "\0"));
```

Comment:
What is the purpose of the null delimiter here? Should this also be an `emplace` instead of an `insert`?

Reply:
We could use `emplace` for performance and safety reasons. The null delimiter is there due to the `header_value` manipulation. We remove `CRLF` characters following the header value from the response and we ensure that the header value is null terminated this way.

Reply:
> we ensure that the redirection value is null terminated this way.

Thanks for clarifying the intent. C++ string do not need to be null delimited when created, you can remove the null delimiter from all your string creations.